[ConstraintSystem] Don't update the work list when merging type variables in addJoinConstraint

hborla opened the pull request:
The missing `/*updateWorkList=*/false` argument turns `mergeEquivalenceClasses` into a very expensive operation because it needs to gather constraints in order to update the worklist. It isn't necessary to activate these constraints anyway because the `conforms to ExpressibleBy*` constraints will be unsolvable at this point.

Resolves: rdar://problem/66807959

hborla committed:
[ConstraintSystem] Don't update the work list when merging type variables
in addJoinConstraint.

hborla commented:
@swift-ci please smoke test

hamishknight commented:
Should we remove the default argument for `updateWorkList`? It looks like there's only 2 callers relying on it, with the majority passing `false`.

hborla commented:
@hamishknight Yeah, I think that's a good idea. That certainly would have prevented me from making this mistake 😅

hborla committed:
[ConstraintSystem] Remove default argument for updateWorkList from
ConstraintSystem::mergeEquivalenceClasses.

Most callers don't use the default, and it's important to consider
the value of this argument for each call to mergeEquivalenceClasses.

hborla commented:
@swift-ci please smoke test
